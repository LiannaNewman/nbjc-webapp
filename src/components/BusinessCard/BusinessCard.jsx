import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import cx from 'classnames';

import {
  Button,
  Card,
  CardHeader,
  CardMedia,
  CardContent,
  Divider,
  Typography,
} from '@material-ui/core';
import StarIcon from '@material-ui/icons/Star';
import BookmarkBorderIcon from '@material-ui/icons/BookmarkBorder';
import ShareIcon from '@material-ui/icons/Share';

import { spaceProps } from '../../types';
import ChipList from '../ChipList';

const styles = {
  root: {
    '& .MuiSvgIcon-root': {
      padding: '6px 6px;',
    },
    maxWidth: 345,
  },
  rating: {
    display: 'block',
    'text-align': 'center',
  },
  cardMedia: {
    width: 350,
    height: 200,
    margin: 'auto',
  },
  // todo: this needs to be a hyrperlink
  location: {
    display: 'flex',
    'flex-direction': 'row',
    marginBottom: 15,
    color: 'rgba(0, 0, 0, 0.6)',
    textDecorationLine: 'underline',
  },

  distance: {
    textAlign: 'right',
    flexGrow: 2,
  },
  address: {
    marginRight: 30,
    flexGrow: 1,
  },
  chipWrapper: {
    'margin-top': '15px',
  },
  filter: {
    margin: '15px 20px 15px 0',
  },
  shareButton: {
    float: 'right',
    padding: '6px 6px',
  },
};

const BusinessCard = ({
  business: {
    name,
    category,
    averageRating,
    imageUrl,
    address,
    distance,
    filters,
  },
  classes,
  overrideClasses,
}) => (
  <Card className={cx(classes.root, overrideClasses.root)} variant="outlined">
    <CardHeader
      avatar={(
        <div color="secondary">
          <StarIcon color="secondary" fontSize="large" />
          <Typography
            variant="body2"
            color="secondary"
            className={classes.rating}
          >
            {averageRating}
          </Typography>
        </div>
      )}
      action={<BookmarkBorderIcon fontSize="large" color="secondary" />}
      title={<Typography variant="h6">{name}</Typography>}
      subheader={<Typography variant="body2">{category}</Typography>}
    />
    <CardMedia
      image={imageUrl}
      // todo make this responsive
      className={classes.cardMedia}
    />
    <CardContent>
      <div className={classes.location}>
        <div className={classes.address}>
          <Typography variant="body1">{address}</Typography>
        </div>
        <div className={classes.distance}>
          <Typography variant="body1">{distance}</Typography>
        </div>
      </div>
      <Divider />
      <div className={classes.chipWrapper}>
        <ChipList chips={filters} />
      </div>
      <div>
        <Button color="secondary">Call Space</Button>
        <Button color="secondary">Write a review</Button>
        <ShareIcon
          color="secondary"
          size="small"
          className={classes.shareButton}
        />
      </div>
    </CardContent>
  </Card>
);

BusinessCard.propTypes = {
  business: PropTypes.shape(spaceProps).isRequired,
  classes: PropTypes.shape({}).isRequired,
  overrideClasses: PropTypes.shape({}),
};

BusinessCard.defaultProps = {
  overrideClasses: {},
};

export default withStyles(styles)(BusinessCard);
