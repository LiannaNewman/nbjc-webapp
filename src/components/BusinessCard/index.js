import BusinessCard from './BusinessCard.jsx';

export default BusinessCard;