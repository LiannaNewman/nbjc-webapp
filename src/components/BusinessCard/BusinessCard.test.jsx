/* eslint-disable no-unused-vars */

import React from 'react';
import { getByText, render, screen } from '@testing-library/react';
import { Share } from '@material-ui/icons';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import BusinessCard, { Tests } from './BusinessCard';
import MockChipList from '../ChipList';

jest.mock('../ChipList', () => () => (<></>));

const {
  convertAddressToGoogleMapsLink,
  formatPhoneNumber,
  Image,
  TopCardContent,
  OrdinalNumber,
  ShareMenu,
  TitleBar,
  AddressRow,
  AddReviewCTA,
  RatingCTA,
  CallPhoneCTA,
  VisitWebsiteCTA,
  CTAs,
  BottomCardContent,
} = Tests;

describe('BusinessCard', () => {
  describe('convertAddressToGoogleMapsLink', () => {
    const googleAPIQuery = 'https://www.google.com/maps/dir/?api=1&destination=';
    it('It should return a link to Google Maps that is URL encoded', () => {
      const googleMapsLink = convertAddressToGoogleMapsLink('1701 Pennsylvania Ave NW Washington, DC DC 20006');
      const stringToMatch = `${googleAPIQuery}1701%20Pennsylvania%20Ave%20NW%20Washington%2C%20DC%20DC%2020006`;
      expect(googleMapsLink).toBe(stringToMatch);
    });
    it('It should return a different link to Google Maps that is URL encoded', () => {
      const googleMapsLink = convertAddressToGoogleMapsLink('651 Florida Ave NW Washington, DC DC 20001');
      const stringToMatch = `${googleAPIQuery}651%20Florida%20Ave%20NW%20Washington%2C%20DC%20DC%2020001`;
      expect(googleMapsLink).toBe(stringToMatch);
    });
  });
  describe('formatPhoneNumber', () => {
    it('It should return a string saying no phone number found if no phone number is passed in', () => {
      const phoneNumber = formatPhoneNumber(undefined);
      expect(phoneNumber).toBe('No number found');
    });
  });
  describe('OrdinalNumber', () => {
    const dummyProps = {
      classes: {},
      count: 1,
    };
    it('It should take an integer value and add a . in front of it', () => {
      render(<OrdinalNumber {...dummyProps} />);
      expect(screen.getByText('1.')).toBeInTheDocument();
    });
  });
  describe('ShareMenu', () => {
    it('It should open up a menu item that has the label share on click', () => {
      render(<ShareMenu classes={{}} />);
      const shareButton = screen.getByTestId('share-button');
      expect(screen.queryByText('Share')).not.toBeInTheDocument();

      userEvent.click(shareButton);
      expect(screen.getByText('Share')).toBeInTheDocument();
    });
  });
  describe('TitleBar', () => {
    const dummyProps = {
      classes: {},
      useDesktop: true,
      count: 5,
      businessName: 'Free Newspapers for $10.99',
    };
    it('It should render a business name', () => {
      render(<TitleBar {...dummyProps} />);
      expect(screen.getByText('Free Newspapers for $10.99')).toBeInTheDocument();
    });
    it('It should render an ordinal number', () => {
      render(<TitleBar {...dummyProps} />);
      expect(screen.getByText('5.')).toBeInTheDocument();
    });
  });
  describe('AddressRow', () => {
    const dummyProps = {
      classes: {},
      address: '1701 Pennsylvania Ave NW Washington, DC DC 20006',
    };
    it('It should have an icon', () => {
      render(<AddressRow {...dummyProps} />);
      const icon = screen.getByTestId('location-icon');
      expect(icon).toBeInTheDocument();
    });
    it('It should have the address label', () => {
      render(<AddressRow {...dummyProps} />);
      const address = screen.getByText('1701 Pennsylvania Ave NW Washington, DC DC 20006');
      expect(address).toBeInTheDocument();
    });
    it('It should have an href value of google maps', () => {
      render(<AddressRow {...dummyProps} />);
      const addressLink = screen.getByRole('link');
      expect(addressLink).toHaveAttribute('href', 'https://www.google.com/maps/dir/?api=1&destination=1701%20Pennsylvania%20Ave%20NW%20Washington%2C%20DC%20DC%2020006');
    });
  });
  describe('AddReviewCTA', () => {
    const dummyProps = {
      classes: {},
      id: 1,
    };
    const renderReviewCTA = () => {
      render(
        <BrowserRouter>
          <AddReviewCTA {...dummyProps} />
        </BrowserRouter>,
      );
    };
    it('It should have a href of /spaces/id/reviews/new', () => {
      renderReviewCTA();
      expect(screen.getByRole('link')).toHaveAttribute('href', '/spaces/1/reviews/new');
    });
    it('It should have a correct label', () => {
      renderReviewCTA();
      expect(screen.getByText('Add Review')).toBeInTheDocument();
    });
    it('It should have an icon', () => {
      renderReviewCTA();
      expect(screen.getByTestId('review-icon')).toBeInTheDocument();
    });
  });
  describe('RatingCTA', () => {

  });
});

/*
const RatingCTA = ({ classes, averageRating }) => {
  const label = 'Rating';
  return (
    <Box className={classes.ctaButtonContainer}>
      <span>{label}</span>
      <StarIcon color="secondary" fontSize="small" />
      <span className={classes.purpleIcon}>{averageRating}</span>
    </Box>
  );
};
*/

/*
test('renders the BusinessCardComponent', () => {
  const props = {
    id: '123',
    name: 'La colombe coffee roasters',
    category: 'Coffee',
    averageRating: '4.5',
    imageUrl:
      'https://s3-media3.fl.yelpcdn.com/bphoto/ONmc2uTWa4W5AAI2irBpXw/o.jpg',
    address: '924 Blagden Alley Way Washington, D.C. 20001',
    distance: '0.03mi',
    filters: [
      {
        name: 'Black Friendly',
      },
      {
        name: 'Inclusive',
      },
      {
        name: 'Black Owned',
      },
      {
        name: 'Gender Neutral Restrooms',
      },
      {
        name: 'Accessible',
      },
      {
        name: 'Queer hangout space',
      },
      {
        name: 'Trans friendly',
      },
      {
        name: 'Queer owned',
      },
    ],
    phoneNumber: '000000000',
    url: 'example.com',
  };
  const { asFragment } = render(<BusinessCard business={props} />);
  expect(asFragment()).toMatchSnapshot();
});
*/
