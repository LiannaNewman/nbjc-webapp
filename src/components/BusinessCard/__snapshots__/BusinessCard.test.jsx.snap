// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the BusinessCardComponent 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root BusinessCard-root-1 MuiPaper-outlined MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar"
      >
        <div
          color="secondary"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeLarge"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
            />
          </svg>
          <p
            class="MuiTypography-root BusinessCard-rating-2 MuiTypography-body2 MuiTypography-colorSecondary"
          >
            4.5
          </p>
        </div>
      </div>
      <div
        class="MuiCardHeader-content"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6"
        >
          La colombe coffee roasters
        </h6>
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          Coffee
        </p>
      </div>
      <div
        class="MuiCardHeader-action"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorSecondary MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2zm0 15l-5-2.18L7 18V5h10v13z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiCardMedia-root BusinessCard-cardMedia-3"
      style="background-image: url(https://s3-media3.fl.yelpcdn.com/bphoto/ONmc2uTWa4W5AAI2irBpXw/o.jpg);"
    />
    <div
      class="MuiCardContent-root"
    >
      <div
        class="BusinessCard-location-4"
      >
        <div
          class="BusinessCard-address-6"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            924 Blagden Alley Way Washington, D.C. 20001
          </p>
        </div>
        <div
          class="BusinessCard-distance-5"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            0.03mi
          </p>
        </div>
      </div>
      <hr
        class="MuiDivider-root"
      />
      <div
        class="BusinessCard-chipWrapper-7"
      />
      <div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Call Space
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Write a review
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root BusinessCard-shareButton-9 MuiSvgIcon-colorSecondary"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
